File: practice/algorithms/graph_theory/snakes_and_ladders/snakes_and_ladders.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <set>
#include <queue>
#include <map>

using namespace std;

int shortest_path(const vector<set<int>> & i_graph)
{
    vector<int> dists(100, 1000000);
    vector<bool> visited(100, false);
    
    dists[0] = 0;
    visited[0] = true;
    
    queue<int> q;
    q.push(0);
    
    while (!q.empty())
    {
        int vertex = q.front();
        q.pop();
        visited[vertex] = true;
        if (vertex == 99)
        {
            return dists[vertex];
        }
        
        set<int>::const_iterator it = i_graph[vertex].begin();
        for (; it != i_graph[vertex].end(); ++it)
        {
            if (!visited[*it])
            {
                dists[*it] = min(dists[*it], dists[vertex] + 1);
                q.push(*it);
            }
        }
    }
    return -1;
}

int main()
{
    int t, num_ladders, num_snakes;
    int s, e;
    cin >> t;

    while (t--)
    {
        map<int, int> ladders, snakes;
        cin >> num_ladders;
        while (num_ladders--)
        {
            cin >> s >> e;
            ladders[s - 1] = e - 1;
        }
        
        cin >> num_snakes;
        while (num_snakes--)
        {
            cin >> s >> e;
            snakes[s - 1] = e - 1;
        }
        
        vector<set<int>> graph(100);
        for (int start = 0; start < 100; ++start)
        {
            for (int end = start + 1; end < start + 7; ++end)
            {
                if (ladders.find(end) != ladders.end())
                {
                    graph[start].insert(ladders[end]);
                }
                if (snakes.find(end) != snakes.end())
                {
                    graph[start].insert(snakes[end]);
                }
            }
            int end = min(start + 6, 99);
            if (ladders.find(end) == ladders.end() && snakes.find(end) == snakes.end())
            {
                graph[start].insert(end);
            }
        }
    
        cout << shortest_path(graph) << endl;
    }
    
    return 0;
}

File: practice/algorithms/graph_theory/snakes_and_ladders/snakes_and_ladders.py
#!/usr/bin/env python

def shortest_path(i_graph):
    dists = [1000000 for _ in range(100)]
    dists[0] = 0
    q = [0]
    visited = set([0])
    while q:
        vertex = q.pop(0)
        visited.add(vertex)
        if vertex == 99:
            return dists[vertex]
        for NODE in i_graph[vertex]:
            if NODE not in visited:
                dists[NODE] = min(dists[NODE], dists[vertex] + 1)
                q.append(NODE)
    return -1

t = int(raw_input().strip())
for _ in range(t):
    num_ladders = int(raw_input().strip())
    ladders = {}
    for __ in range(num_ladders):
        s, e = map(int, raw_input().strip().split())
        ladders[s - 1] = e - 1
    
    num_snakes = int(raw_input().strip())
    snakes = {}
    for __ in range(num_snakes):
        s, e = map(int, raw_input().strip().split())
        snakes[s - 1] = e - 1
        
    graph = [set() for _ in range(100)]
    for START in range(100):
        for END in range(START + 1, START + 7):
            if END < 100:
                if END in ladders.keys():
                    graph[START].add(ladders[END])
                elif END in snakes.keys():
                    graph[START].add(snakes[END])
        END = min(START + 6, 99)
        if END not in ladders.keys() and END not in snakes.keys():
            graph[START].add(END)
                
    print shortest_path(graph)

